README.md
# clickable-prototype

> Lightweight React Library for creating clickable Prototypes

[![NPM](https://img.shields.io/npm/v/clickable-prototype.svg)](https://www.npmjs.com/package/clickable-prototype) [![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com) [![Build Status](https://travis-ci.org/pinussilvestrus/clickable-prototype.svg?branch=master)](https://travis-ci.org/pinussilvestrus/clickable-prototype) ![npm bundle size](https://img.shields.io/bundlephobia/min/clickable-prototype)

[**Demo Page**](https://clickable-prototype.netlify.com/)

## Install

```bash
npm install --save clickable-prototype
```

## Usage

```jsx
import React, { Component } from 'react'

import { View, HitBox, ViewContainer } from 'clickable-prototype'

import view1 from './views/view1.png'
import view2 from './views/view2.png'

export default class App extends Component {
  render () {
    return (
      <div>
        <h1><span>clickable-prototype</span> Demo</h1>
        <ViewContainer defaultView='view1' className='container'>
          <View
            key='view1'
            className='custom-view'
            screen={view1}
            width='1220px'
            height='630px'>
            <HitBox
              position={{y: 380, x: 477, width: 280, height: 60}}
              to='view2' />
          </View>
          <View
            key='view2'
            className='custom-view'
            screen={view2}
            width='1220px'
            height='630px'>
            <HitBox
              position={{y: 490, x: 477, width: 280, height: 60}}
              to='view1' />
          </View>
        </ViewContainer>
      </div>
    )
  }
}

```

## Development Setup

To set up this project, first clone the repository
```bash
$ git clone https://github.com/pinussilvestrus/clickable prototype
```

Change your working directory into the project directory
```bash
$ cd clickable-prototype
```

Install all dependencies in the library and the [example](./example)

```bash
$ npm install
$ cd example && npm install
```

Run the development server(s)

```bash
$ npm run dev
```

## License

MIT © [pinussilvestrus](https://github.com/pinussilvestrus)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased Changes

* /

## v0.4.0

* `CHORE`: bump dependencies

## v0.3.0

* `FEAT`: ability to declare width / height on `View` component ([#6](https://github.com/pinussilvestrus/clickable-prototype/issues/6))
* `FEAT`: ability to declare className on `View` component ([`d24388ec`](https://github.com/pinussilvestrus/clickable-prototype/commit/d24388ec83e03e3b24884950ccbbd9cda6c687e9))

## v0.2.0

* `FIX`: fix `npm audit` errors

## v0.1.0

Initial Release

src/test.js
import View from './'

describe('View', () => {
  it('is truthy', () => {
    expect(View).toBeTruthy()
  })
})


src/components/__tests__/ViewSpec.js
import React from 'react'
import { shallow } from 'enzyme'

import View from '../View'

describe('View', () => {
  it('should render', () => {
    shallow(<View />)
  })
})


src/components/__tests__/HitBoxSpec.js
import React from 'react'
import { shallow } from 'enzyme'

import HitBox from '../HitBox'

describe('HitBox', () => {
  it('should render', () => {
    // given
    const position = { x: 0, y: 0, width: 0, height: 0 }

    // then
    shallow(<HitBox position={position} />)
  })

  it('should crash without position', () => {
    expect(() => {
      shallow(<HitBox />)
    }).toThrow()
  })
})


rollup.config.js
import babel from 'rollup-plugin-babel'
import commonjs from '@rollup/plugin-commonjs'
import external from 'rollup-plugin-peer-deps-external'
import postcss from 'rollup-plugin-postcss'
import resolve from '@rollup/plugin-node-resolve'
import url from '@rollup/plugin-url'
import svgr from '@svgr/rollup'

import pkg from './package.json'

export default {
  input: 'src/index.js',
  output: [
    {
      file: pkg.main,
      format: 'cjs',
      sourcemap: true
    },
    {
      file: pkg.module,
      format: 'es',
      sourcemap: true
    }
  ],
  plugins: [
    external(),
    postcss({
      modules: true
    }),
    url(),
    svgr(),
    babel({
      exclude: 'node_modules/**',
      plugins: [
        '@babel/plugin-proposal-class-properties'
      ]
    }),
    resolve(),
    commonjs()
  ]
}


src/components/View.js
import React, { Component } from 'react'
import PropTypes from 'prop-types'

import styles from '../styles.css'

export default class View extends Component {
  static propTypes = {
    screen: PropTypes.string,
    children: PropTypes.object,
    onSwitchView: PropTypes.func,
    visible: PropTypes.bool,
    width: PropTypes.string,
    height: PropTypes.string,
    className: PropTypes.string
  }

  handleAreaClick = (area) => {
    const {
      onSwitchView
    } = this.props

    onSwitchView(area.props.to)
  }

  render() {
    const {
      screen,
      className,
      children,
      visible,
      width,
      height
    } = this.props

    const handleAreaClick = this.handleAreaClick

    const areaDiv = React.Children.map(children, child => {
      return (
        <div onClick={() => handleAreaClick(child)}>{child}</div>
      )
    })

    return (
      <div
        className={`${styles.view} ${className}`}
        style={{
          visibility: visible ? 'visible' : 'hidden',
          width,
          height
        }}>
        <img src={screen} />
        { areaDiv }
      </div>
    )
  }
}


src/components/ViewContainer.js
import React, { Component } from 'react'
import PropTypes from 'prop-types'

export default class ViewContainer extends Component {
  static propTypes = {
    defaultView: PropTypes.string,
    children: PropTypes.array,
    className: PropTypes.string
  }

  state = {
    activeView: null
  }

  componentDidMount() {
    this.setState({
      activeView: this.props.defaultView
    })
  }

  handleSwitchView = (activeView) => {
    this.setState({
      activeView
    })
  }

  render() {
    const {
      activeView
    } = this.state

    const onSwitchView = this.handleSwitchView

    const children = React.Children.map(this.props.children, child => {
      return React.cloneElement(child, {
        onSwitchView,
        visible: activeView === child.key
      })
    })

    return (
      <div
        className={this.props.className}>
        {children}
      </div>
    )
  }
}


example/src/App.js
import React, { Component } from 'react'

import { View, HitBox, ViewContainer } from 'clickable-prototype'

import view1 from './views/view1.png'
import view2 from './views/view2.png'

export default class App extends Component {
  render () {
    return (
      <div>
        <h1><span>clickable-prototype</span> Demo</h1>
        <ViewContainer defaultView='view1' className='container'>
          <View
            key='view1'
            className='custom-view'
            screen={view1}
            width='1220px'
            height='630px'>
            <HitBox
              position={{y: 380, x: 477, width: 280, height: 60}}
              to='view2' />
          </View>
          <View
            key='view2'
            className='custom-view'
            screen={view2}
            width='1220px'
            height='630px'>
            <HitBox
              position={{y: 490, x: 477, width: 280, height: 60}}
              to='view1' />
          </View>
        </ViewContainer>
      </div>
    )
  }
}
